examples/Basic/gui_table.py

```diff
@@ -0,0 +1,94 @@
+"""
+This example demonstrates how to make a graphical interface, and uses
+a random number generator to simulate data so that it does not require
+an instrument to use. In particular, this example shows how to display
+data in tabular format.
+
+Run the program by changing to the directory containing this file and calling:
+
+python gui_table.py
+
+"""
+
+import sys
+import random
+from time import sleep
+from pymeasure.experiment import Procedure, IntegerParameter, Parameter, FloatParameter
+from pymeasure.experiment import Results, unique_filename
+from pymeasure.display.Qt import QtWidgets
+from pymeasure.display.windows import ManagedWindowBase
+from pymeasure.display.widgets import TableWidget, LogWidget
+
+import logging
+log = logging.getLogger('')
+log.addHandler(logging.NullHandler())
+
+
+class TestProcedure(Procedure):
+
+    iterations = IntegerParameter('Loop Iterations', default=10)
+    delay = FloatParameter('Delay Time', units='s', default=0.2)
+    seed = Parameter('Random Seed', default='12345')
+
+    DATA_COLUMNS = ['Iteration', 'Random Number']
+
+    def startup(self):
+        log.info("Setting up random number generator")
+        random.seed(self.seed)
+
+    def execute(self):
+        log.info("Starting to generate numbers")
+        for i in range(self.iterations):
+            data = {
+                'Iteration': i,
+                'Random Number': random.random()
+            }
+            log.debug("Produced numbers: %s" % data)
+            self.emit('results', data)
+            self.emit('progress', 100 * i / self.iterations)
+            sleep(self.delay)
+            if self.should_stop():
+                log.warning("Catch stop command in procedure")
+                break
+
+    def shutdown(self):
+        log.info("Finished")
+
+
+class MainWindow(ManagedWindowBase):
+
+    def __init__(self):
+        widget_list = (TableWidget("Experiment Table",
+                                   TestProcedure.DATA_COLUMNS,
```

Comment:
I think this idea of index column is nice. Just to double check, I understand this correctly:
- it is valid only  when using by column representation
- the idea is to have only one index column (iterations, in the example) to avoid repetition of it.

@CasperSchippers, is my understanding correct?

Reply:
Indeed, that is exactly what I meant

Reply:
I have tried to implement the indexing in commit https://github.com/pymeasure/pymeasure/commit/bf72c1f706b02bbb2621a7e774ff916738c70316

I really would love to have your comments, because functionally speaking is ok, but I feel code is more complex that it should be.

Reply:
I played around with this implementation and I really like it, well done!! Also the fact that you can change the indexing column and switch between the "by column" or "by row" mode during runtime is really nice. I need to take a bit more time to look at the code to see if it is more complex than it should be.

Reply:
I've had a look; at first it seemed indeed a bit complex to me (especially since I'm not familiar with the abstract models), but I did not really see a way to make it simpler. I do think that the added functionality outweighs the complexity in this case, so I'm fine with it.